### PR TITLE
SONARJS-802 Rule: Arithmetic operations should not always result in NaN

### PR DESCRIPTION
**Pull request by pynicolas**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    public void check(Constraint leftConstraint, Constraint rightConstraint, Tree leftOperand, Tree rightOperand, Tree operator) {
+      if (isNumberAndUndefined(rightConstraint, leftConstraint)) {
+        raiseIssue(leftOperand, operator, rightOperand);
+      } else if (isNumberAndUndefined(leftConstraint, rightConstraint)) {
+        raiseIssue(rightOperand, operator, leftOperand);
+      }
+    }
+
+    private boolean isNumberAndUndefined(Constraint constraint1, Constraint constraint2) {
+      return constraint1.isStricterOrEqualTo(NUMBER_OR_BOOLEAN_OR_UNDEFINED)
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java]: @pynicolas Why `NUMBER_OR_BOOLEAN_OR_UNDEFINED` and not `NUMBER_OR_BOOLEAN`. btw It's not tested

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        && constraint2.isStricterOrEqualTo(Constraint.UNDEFINED);
+    }
+
+  }
+
+  private class MinusChecker implements BinaryOperationChecker {
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java]: @pynicolas can we make these classes `static`?

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+  }
+
+  private void raiseIssue(Tree operand, Tree... secondaryLocations) {
+    Tree expression = ((JavaScriptTree) operand).getParent();
+    Symbol operandSymbol = null;
+    if (operand.is(Kind.IDENTIFIER_REFERENCE)) {
+      IdentifierTree identifier = (IdentifierTree) operand;
+      operandSymbol = identifier.symbol();
+    }
+    if (!expressionsWithIssues.contains(expression) && !symbolsWithIssues.contains(operandSymbol)) {
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java]: @pynicolas no need for `!expressionsWithIssues.contains(expression)` as you don't raise 2 issues for one expression (at least in test)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java (hunk 2 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java]: @vilchik-elena No, we can't, they use `raiseIssue` which cannot be `static`.

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3757.html [hunk 4]
```diff
@@ -0,0 +1,13 @@
+<p>The result of an expression with an arithmetic operator <code>/, *, %, ++, --, -, +=, -=, *=, /=, %=, +</code> or unary operator <code>+, -</code>
+when at least one operand is <code>Object</code> or <code>Undefined</code> will be always a <code>NaN</code> (Not a Number).</p>
+<h2>Noncompliant Code Example</h2>
+<pre>
+x = [1, 2];
+var y = x / 4;  //Non-compliant [BUG] &gt; "y" == NaN
```
Comment (pynicolas) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3757.html]: @vilchik-elena done in 9262a3a.

File: javascript-checks/src/test/resources/checks/ArithmeticOperationReturningNan.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
+function f1() {
+  42 - 7;
+  42 - []; // Noncompliant {{Change the expression using this operand so that it can't evaluate to "NaN" (Not a Number).}}
```
Comment (pynicolas) [inline javascript-checks/src/test/resources/checks/ArithmeticOperationReturningNan.js]: @vilchik-elena Changed in 9262a3a.

File: javascript-checks/src/test/resources/checks/ArithmeticOperationReturningNan.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+       42;
+//S    ^^ 1
+
+  [] / 42; // Noncompliant
+  [] * 42; // Noncompliant
+  [] % 42; // Noncompliant
+
+  var array1 = [];
+  array1++; // Noncompliant
+//^^^^^^
```
Comment (pynicolas) [inline javascript-checks/src/test/resources/checks/ArithmeticOperationReturningNan.js]: @vilchik-elena Changed in 9262a3a

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java (hunk 3 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java]: @vilchik-elena  Right. I removed that in 9262a3a.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java (hunk 1 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArithmeticOperationReturningNanCheck.java]: @vilchik-elena Now, it's tested (9262a3a).

**Pull request merged**